### ISSUE TITLE
staticcheck(SA4023): invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by jhawk28** [golangci/golangci-lint#6746]
### Welcome

- [x] Yes, I'm using a binary release within 2 latest [releases](https://github.com/golangci/golangci-lint/releases). Only such installations are supported.
- [x] Yes, I've searched similar [issues on GitHub](https://github.com/golangci/golangci-lint/issues) and didn't find any.
- [x] Yes, I've read the `typecheck` section of the [FAQ](https://golangci-lint.run/docs/welcome/faq/#why-do-you-have-typecheck-errors).
- [x] Yes, I've tried with the standalone [linter](https://golangci-lint.run/docs/linters/) if available (e.g., gocritic, go vet, etc.).
- [x] I agree to follow this project's [Code of Conduct](https://github.com/golangci/golangci-lint?tab=coc-ov-file)

### How did you install golangci-lint?

go install

### Description of the problem

Using the latest version of golangci-lint (go install github.com/golangci/golangci-lint/v2/cmd/golangci-lint@latest). 

Fails with following stacktrace:
```
Panic: SA4023: package "sui" (isInitialPkg: true, needAnalyzeSource: true): runtime error: invalid memory address or nil pointer dereference: goroutine 43176 [running]:
runtime/debug.Stack()
	/usr/local/go/src/runtime/debug/stack.go:26 +0x64
github.com/golangci/golangci-lint/v2/pkg/goanalysis.(*action).analyzeSafe.func1()
	/Users/joshua/go/pkg/mod/github.com/golangci/golangci-lint/v2@v2.13.1/pkg/goanalysis/runner_action.go:55 +0x1d8
panic({0x1044b87c8?, 0x10479a280?})
	/usr/local/go/src/runtime/panic.go:859 +0x120
honnef.co/go/tools/go/ir.(*Function).Source(...)
	/Users/joshua/go/pkg/mod/honnef.co/go/tools@v0.8.0/go/ir/ssa.go:1979
honnef.co/go/tools/go/ir/irutil.isTrivial(0x6e8bb31d5ce0?, 0x6e8bc337f2c0?)
	/Users/joshua/go/pkg/mod/honnef.co/go/tools@v0.8.0/go/ir/irutil/stub.go:49 +0x18
honnef.co/go/tools/go/ir/irutil.IsTrivial(...)
	/Users/joshua/go/pkg/mod/honnef.co/go/tools@v0.8.0/go/ir/irutil/stub.go:45
honnef.co/go/tools/staticcheck/sa4023.run(0x6e8bb31d5ce0)
	/Users/joshua/go/pkg/mod/honnef.co/go/tools@v0.8.0/staticcheck/sa4023/sa4023.go:184 +0x600
github.com/golangci/golangci-lint/v2/pkg/goanalysis.(*action).analyze.func3(...)
	/Users/joshua/go/pkg/mod/github.com/golangci/golangci-lint/v2@v2.13.1/pkg/goanalysis/runner_checker.go:180
github.com/golangci/golangci-lint/v2/pkg/goanalysis.(*action).analyze(0x6e8bb69eee40)
	/Users/joshua/go/pkg/mod/github.com/golangci/golangci-lint/v2@v2.13.1/pkg/goanalysis/runner_checker.go:206 +0x9d8
github.com/golangci/golangci-lint/v2/pkg/timeutils.(*Stopwatch).TrackStage(0x6e8bb50c4b40, {0x1030b03c0, 0x6}, 0x6e8bb5e58700)
	/Users/joshua/go/pkg/mod/github.com/golangci/golangci-lint/v2@v2.13.1/pkg/timeutils/stopwatch.go:111 +0x44
github.com/golangci/golangci-lint/v2/pkg/goanalysis.(*action).analyzeSafe(0x6e8bb69eee40?)
	/Users/joshua/go/pkg/mod/github.com/golangci/golangci-lint/v2@v2.13.1/pkg/goanalysis/runner_action.go:59 +0x64
github.com/golangci/golangci-lint/v2/pkg/goanalysis.(*loadingPackage).analyze.func2()
	/Users/joshua/go/pkg/mod/github.com/golangci/golangci-lint/v2@v2.13.1/pkg/goanalysis/runner_loadingpackage.go:107 +0x6c
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/Users/joshua/go/pkg/mod/golang.org/x/sync@v0.22.0/errgroup/errgroup.go:93 +0x4c
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 459
	/Users/joshua/go/pkg/mod/golang.org/x/sync@v0.22.0/errgroup/errgroup.go:78 +0x98
```

### Version of golangci-lint

golangci-lint has version 2.13.1 built with go1.27.0 from (unknown, modified: ?, mod sum: "h1:RuM4OcluM4xFQcGuRE6R7jA33pqxK/W1EsBxpugdZjg=") on (unknown)

### Configuration

<details>

```console
version: "2"
run:
  tests: false
  timeout: 5m
  modules-download-mode: readonly

linters:
  default: standard
  exclusions:
    rules:
      - linters:
          - staticcheck
        text: "QF1008:" # Omit embedded fields from selector expression.
      - linters:
          - staticcheck
        text: "QF1001:" # could apply De Morgan's law
```

</details>

### Go environment

<details>

```console
go version go1.27.0 darwin/arm64
```

</details>

### Verbose output of running

<details>

```console
INFO golangci-lint has version 2.13.1 built with go1.27.0 from (unknown, modified: ?, mod sum: "h1:RuM4OcluM4xFQcGuRE6R7jA33pqxK/W1EsBxpugdZjg=") on (unknown) 
INFO [config_reader] Config search paths: [./ /Users/joshua/repos/node /Users/joshua/repos /Users/joshua /Users /] 
INFO [config_reader] Used config file .golangci.yml 
INFO [config_reader] Module name "github.com/quiknode-labs/node/v4" 
INFO [goenv] Read go env for 15.686459ms: map[string]string{"GOCACHE":"/Users/joshua/Library/Caches/go-build", "GOROOT":"/usr/local/go"} 
INFO [lintersdb] Active 5 linters: [errcheck govet ineffassign staticcheck unused] 
INFO [loader] Go packages loading at mode 8767 (deps|files|imports|types_sizes|compiled_files|exports_file|name) took 1.226075667s 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 17.248209ms 
INFO [linters_context/goanalysis] analyzers took 6.808655291s with top 10 stages: nilness: 580.686999ms, unused: 552.833255ms, buildir: 409.546679ms, S1008: 199.469463ms, directives: 193.725506ms, fact_purity: 190.381198ms, printf: 175.319892ms, S1038: 153.051365ms, typeindex: 146.696328ms, fact_deprecated: 145.191254ms 
ERRO [runner] Panic: SA4023: package "sui" (isInitialPkg: true, needAnalyzeSource: true): runtime error: invalid memory address or nil pointer dereference: goroutine 45299 [running]:
runtime/debug.Stack()
	/usr/local/go/src/runtime/debug/stack.go:26 +0x64
github.com/golangci/golangci-lint/v2/pkg/goanalysis.(*action).analyzeSafe.func1()
	/Users/joshua/go/pkg/mod/github.com/golangci/golangci-lint/v2@v2.13.1/pkg/goanalysis/runner_action.go:55 +0x1d8
panic({0x102e347c8?, 0x103116280?})
	/usr/local/go/src/runtime/panic.go:859 +0x120
honnef.co/go/tools/go/ir.(*Function).Source(...)
	/Users/joshua/go/pkg/mod/honnef.co/go/tools@v0.8.0/go/ir/ssa.go:1979
honnef.co/go/tools/go/ir/irutil.isTrivial(0x215d0c4f9c00?, 0x215d00b21d40?)
	/Users/joshua/go/pkg/mod/honnef.co/go/tools@v0.8.0/go/ir/irutil/stub.go:49 +0x18
honnef.co/go/tools/go/ir/irutil.IsTrivial(...)
	/Users/joshua/go/pkg/mod/honnef.co/go/tools@v0.8.0/go/ir/irutil/stub.go:45
honnef.co/go/tools/staticcheck/sa4023.run(0x215d0c4f9c00)
	/Users/joshua/go/pkg/mod/honnef.co/go/tools@v0.8.0/staticcheck/sa4023/sa4023.go:184 +0x600
github.com/golangci/golangci-lint/v2/pkg/goanalysis.(*action).analyze.func3(...)
	/Users/joshua/go/pkg/mod/github.com/golangci/golangci-lint/v2@v2.13.1/pkg/goanalysis/runner_checker.go:180
github.com/golangci/golangci-lint/v2/pkg/goanalysis.(*action).analyze(0x215cff13ce40)
	/Users/joshua/go/pkg/mod/github.com/golangci/golangci-lint/v2@v2.13.1/pkg/goanalysis/runner_checker.go:206 +0x9d8
github.com/golangci/golangci-lint/v2/pkg/timeutils.(*Stopwatch).TrackStage(0x215cfc2f4140, {0x101a2c3c0, 0x6}, 0x215cfebfe700)
	/Users/joshua/go/pkg/mod/github.com/golangci/golangci-lint/v2@v2.13.1/pkg/timeutils/stopwatch.go:111 +0x44
github.com/golangci/golangci-lint/v2/pkg/goanalysis.(*action).analyzeSafe(0x215cff13ce40?)
	/Users/joshua/go/pkg/mod/github.com/golangci/golangci-lint/v2@v2.13.1/pkg/goanalysis/runner_action.go:59 +0x64
github.com/golangci/golangci-lint/v2/pkg/goanalysis.(*loadingPackage).analyze.func2()
	/Users/joshua/go/pkg/mod/github.com/golangci/golangci-lint/v2@v2.13.1/pkg/goanalysis/runner_loadingpackage.go:107 +0x6c
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/Users/joshua/go/pkg/mod/golang.org/x/sync@v0.22.0/errgroup/errgroup.go:93 +0x4c
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 2712
	/Users/joshua/go/pkg/mod/golang.org/x/sync@v0.22.0/errgroup/errgroup.go:78 +0x98 
WARN [runner] Can't run linter goanalysis_metalinter: goanalysis_metalinter: SA4023: package "sui" (isInitialPkg: true, needAnalyzeSource: true): runtime error: invalid memory address or nil pointer dereference 
INFO [runner/exclusion_rules] Skipped 0 issues by rules: [Text: "QF1008:", Linters: "staticcheck"] 
INFO [runner/exclusion_rules] Skipped 0 issues by rules: [Text: "QF1001:", Linters: "staticcheck"] 
INFO [runner] processing took 4.626µs with stages: exclusion_rules: 3.666µs, max_same_issues: 250ns, exclusion_paths: 167ns, cgo: 84ns, sort_results: 83ns, max_from_linter: 83ns, severity-rules: 42ns, max_per_file_from_linter: 42ns, diff: 42ns, filename_unadjuster: 42ns, path_absoluter: 42ns, uniq_by_line: 42ns, nolint_filter: 41ns, path_shortener: 0s, fixer: 0s, path_relativity: 0s, invalid_issue: 0s, source_code: 0s, path_prettifier: 0s, generated_file_filter: 0s 
INFO [runner] linters took 1.663850708s with stages: goanalysis_metalinter: 1.66381275s 
ERRO Running error: can't run linter goanalysis_metalinter
goanalysis_metalinter: SA4023: package "sui" (isInitialPkg: true, needAnalyzeSource: true): runtime error: invalid memory address or nil pointer dereference 
INFO Memory: 31 samples, avg is 167.7MB, max is 706.5MB 
INFO Execution took 2.923673958s 
```

</details>

### A minimal reproducible example or link to a public repository

<details>

```go
// add your code here
```

</details>

### Validation

- [x] Yes, I've included all the information above (example, version, config, etc.).

### Supporter

- [ ] I am a [sponsor/backer](https://donate.golangci.org) of this project.

**Comment by jhawk28** [comment]
verified that 2.12.2 works.

**Comment by ldez** [comment]
Hello,

can you provide a reproducible as require by the issue template?

---

Workaround:

```yml
version: "2"
linters:
  settings:
    staticcheck:
      checks:
        # default checks
        - "all"
        - "-ST1000"
        - "-ST1003"
        - "-ST1016"
        - "-ST1020"
        - "-ST1021"
        - "-ST1022"
        # Workaround for https://github.com/golangci/golangci-lint/issues/6733 and https://github.com/golangci/golangci-lint/issues/6732
        - "-SA4023"
```

**Comment by jhawk28** [comment]
the workaround works.

**Comment by ldez** [comment]
This a workaround not a fix, if you want a real fix you should provide a reproducible example.

**Comment by jhawk28** [comment]
ack. I closed because it looked like there were 2 other examples...

**Comment by ldez** [comment]
> I closed because it looked like there were 2 other examples...

There is no example of this bug, the previous bugs related to `SA4023` have been fixed in golangci-lint v2.13.1.

So, can you provide a link to a repo or a reproducible code example?

**Comment by jhawk28** [comment]
Not currently able to point to a public repo with a reproducible code example.

It looks like [this](https://github.com/dominikh/go-tools/blob/master/go/ir/irutil/stub.go#L11) is the code that breaks. adding a nil check fixes:
```
if fn == nil || fn.Source() == nil {
	// External functions have to be assumed to not be stubs.
	return false
}
```

**Comment by jhawk28** [comment]
This is the code that causes the issue:

```
subMsg := elementValue.Message()
if subMsg == nil {
	continue
}
```

This is the definition of the .Message() from protoreflect:
```
// Message returns v as a [Message] and panics if the type is not a [Message].
func (v Value) Message() Message {
	switch vi := v.getIface().(type) {
	case Message:
		return vi
	default:
		panic(v.panicMessage("message"))
	}
}

```

**Comment by ldez** [comment]
I tried to create a reproducible example based on your snippet but I have no panic.


<details>

```go
package main

type Value struct{}

func (v *Value) getIface() any {
	return nil
}

func (v *Value) panicMessage(s string) any {
	return nil
}

type Message map[any]any

// Message returns v as a [Message] and panics if the type is not a [Message].
func (v Value) Message() Message {
	switch vi := v.getIface().(type) {
	case Message:
		return vi
	default:
		panic(v.panicMessage("message"))
	}
}

func main() {
	elementValue := Value{}

	for i := 0; i < 2; i++ {
		subMsg := elementValue.Message()
		if subMsg == nil {
			continue
		}
	}
}
```

</details>

**Comment by ldez** [comment]
I also tried this:

```go
package main

import "google.golang.org/protobuf/reflect/protoreflect"

func main() {
	elementValue := protoreflect.Value{}

	for i := 0; i < 2; i++ {
		subMsg := elementValue.Message()
		if subMsg == nil {
			continue
		}
	}
}
```

and no panic.

**Comment by jhawk28** [comment]
its definitely messy. I can verify that removing the `if subMsg == nil {}` block lets it pass. I'm not certain we will be able to get to a simple reproduction. My code is parsing a protobuf [descriptor](https://github.com/MystenLabs/sui-apis/blob/main/proto/sui/rpc/v2/ledger_service.proto#L105), using reflection to get the message description, then it gets the field and walks down to get the value.

**Comment by sdassow** [comment]
We have this issue in Fyne, and can be seen here: https://github.com/fyne-io/fyne/actions/runs/32758537874/job/97533792515?pr=6499

Isolated code to reproduce:
```go
package main

import (
	"log"

	"example/internal"
)

func main() {
	if x := internal.GetTheFoo(); x != nil {
		log.Printf("foo=%v", x)
	}
}
```

```go
package internal

type Foo interface {
	Get() int
}

type foo struct {}

func (f *foo) Get() int {
	return 42
}

var thefoo = &foo{}

func GetTheFoo() Foo {
	return thefoo
}
```

**Comment by ldez** [comment]
Thank you, I should be able to analyze the problem now.

**Comment by ldez** [comment]
I don't reproduce the panic with your snippet.

Your CI is also not panicing.

**Comment by sdassow** [comment]
Oh, my bad. I pulled the trigger too fast on SA4023, which definitely has issues.

**Comment by ldez** [comment]
Your problem is already fixed.
I will create a bugfix release soon (probably tomorrow because I'm waiting for the merge of another bugfix)

**Comment by shoham-utila** [comment]
We hit the same SA4023 nil-deref on golangci-lint **v2.13.2** (staticcheck **v0.8.1**).

The snippets already tried here do not panic because they call `Message()` on a `protoreflect.Value` you already have. That path is a normal SA4023 finding (`this comparison is never true`). The panic only happens when `Message()` is called on an **element of `protopath.Values.Values`** (the type `protorange.Range` callbacks receive). A locally defined struct with the same shape also does **not** panic; it has to be the real `google.golang.org/protobuf/reflect/protopath.Values`.

That matches the original report (`elementValue.Message(); if subMsg == nil`) once `elementValue` comes from `protopath.Values`, not from a bare `protoreflect.Value`.

### Minimal repro

`go.mod`:

```go
module example.com/sa4023repro

go 1.27.0

require google.golang.org/protobuf v1.36.12
```

`walk.go`:

```go
package sa4023repro

import "google.golang.org/protobuf/reflect/protopath"

func Walk(v protopath.Values) bool {
	return v.Values[0].Message() == nil
}
```

```bash
staticcheck -checks SA4023 .
# or
golangci-lint run --default=none -Estaticcheck .
```

Reproduced with:

- `go version go1.27.0 darwin/arm64`
- `staticcheck 2026.2.1 (0.8.1)`
- `golangci-lint has version 2.13.2 built with go1.27.0`

Standalone `staticcheck` panics too, so this is upstream, not a golangci-lint runner issue.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x30]

honnef.co/go/tools/go/ir.(*Function).Source(...)
	honnef.co/go/tools@v0.8.1/go/ir/ssa.go:1979
honnef.co/go/tools/go/ir/irutil.isTrivial(...)
	honnef.co/go/tools@v0.8.1/go/ir/irutil/stub.go:49
honnef.co/go/tools/go/ir/irutil.IsTrivial(...)
	honnef.co/go/tools@v0.8.1/go/ir/irutil/stub.go:45
honnef.co/go/tools/staticcheck/sa4023.run(...)
	honnef.co/go/tools@v0.8.1/staticcheck/sa4023/sa4023.go:192
```

golangci-lint wraps the same panic as:

```
Panic: SA4023: package "sa4023repro" (isInitialPkg: true, needAnalyzeSource: true): runtime error: invalid memory address or nil pointer dereference
```

### Nearby snippets that do **not** panic

```go
// SA4023 reports "this comparison is never true" - no crash
func Walk(v protoreflect.Value) bool {
	return v.Message() == nil
}

// no nil comparison — no crash
func Walk(v protopath.Values) {
	_ = v.Values[0].Message()
}

// same shape as protopath.Values, but a local type - no crash (SA4023 finding only)
type Values struct {
	Values []protoreflect.Value
}

func Walk(v Values) bool {
	return v.Values[0].Message() == nil
}
```

`isTrivial` in v0.8.1 is:

```go
func isTrivial(fn *ir.Function, seen map[*ir.Function]struct{}) bool {
	if fn.Source() == nil { // panics when fn itself is nil
```

A `fn == nil` guard before `fn.Source()` would stop the crash (same idea as in the earlier comment).
